stock_picking_package_preparation_line/models/stock_picking_package_preparation_line.py

```diff
@@ -0,0 +1,280 @@
+        help="If you change this quantity for a 'ready' picking, the system "
+             "will not generate a back order, but will just deliver the new "
+             "quantity")
+    product_uom = fields.Many2one('product.uom')
+    lot_id = fields.Many2one('stock.production.lot', 'Lot')
+    sequence = fields.Integer(default=10)
+    note = fields.Text()
+
+    @api.multi
+    def write(self, values):
+        super(StockPickingPackagePreparationLine, self).write(values)
+        if 'product_uom_qty' in values and self.move_id:
+            if self.move_id.product_uom_qty != values['product_uom_qty']:
+                # perform a new reservation with the new quantity
+                move_context = self.move_id.with_context(
+                    skip_update_line_ids=True)
+                move_context.picking_id.do_unreserve()
+                move_context.product_uom_qty = values['product_uom_qty']
+                move_context.picking_id.action_assign()
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        if self.product_id:
+            name = self.product_id.name_get()
+            if name:
+                self.name = name[0][1]
+            self.product_uom = self.product_id.uom_id.id
+
+    def _prepare_lines_from_pickings(self, picking_ids):
+        lines = []
+        if not picking_ids:
+            return lines
+        picking_model = self.env['stock.picking']
+        for picking in picking_model.browse(picking_ids):
+            for move_line in picking.move_lines:
+                # ----- If stock move is cancel, don't create package
+                #       preparation line
+                if move_line.state == 'cancel':
+                    continue
+                # ----- search if the move is related with a
+                #       PackagePreparationLine, yet. If not, create a new line
+                if not self.search([('move_id', '=', move_line.id)],
+                                   count=True):
+                    lines.append({
+                        'move_id': move_line.id,
+                        'name': move_line.name,
+                        'product_id': move_line.product_id.id,
+                        'product_uom_qty': move_line.product_uom_qty,
+                        'product_uom': move_line.product_uom.id,
+                        'lot_id': move_line.restrict_lot_id.id
+                        if move_line.restrict_lot_id else False,
+                        })
+        return lines
+
+    @api.multi
+    def get_move_data(self):
+        self.ensure_one()
+        return {
+            'name': self.name,
+            'product_id': self.product_id.id,
+            'product_uom_qty': self.product_uom_qty,
+            'product_uom': self.product_uom.id,
+            'restrict_lot_id': self.lot_id.id
+            if self.lot_id else False,
+            }
+
+
+class StockPickingPackagePreparation(models.Model):
```

Comment:
Please either move this into its own file (preferred), or name the file `stock_picking_package_preparation.py`